### PR TITLE
Add logic to clean up jobs that are finished or stuck

### PR DESCRIPTION
**Pull request by artem-shelkovnikov**
## Closes https://github.com/elastic/enterprise-search-team/issues/1980

This PR aims to clean up the jobs that are no longer relevant for the connector. The following scenarios are covered:

- Job has finished and results queue is empty for the job
- Job hasn't had any updates in last X minutes (see the JobWatcher class for specific numbers)

Once the following scenarios are met, the job is either just removed from the storage, or (if it's stuck) failed and then removed from the storage.

#### Pre-Review Checklist
- [x] Covered the changes with automated tests
- [x] Tested the changes locally
- [x] Added a label for each target release version (example: `v7.13.2`, `v7.14.0`, `v8.0.0`)

## For Elastic Internal Use Only
- [x] [Manual test steps](https://github.com/elastic/connectors/blob/main/docs/INTERNAL.md#minimal-manual-tests)  are successful

### CODE REVIEW

File: lib/connectors_async/job_watcher.rb [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+    def watch
+      raise AlreadyWatchingError.new('Already watching!') unless @is_watching.make_true
+      ConnectorsShared::Logger.info('Watching after jobs {•̃_•̃}')
+
+      @pool.post do
+        loop do
+          run!
+        rescue StandardError => e
+          ConnectorsShared::Logger.error("An error occurred while watching over jobs: #{e.message}")
```
Comment (seanstory) [inline lib/connectors_async/job_watcher.rb:38]: Let's use `ConnectorsShared::ExceptionTracking` instead, so that we can get the stack trace too

File: spec/connectors_async/job_watcher_spec.rb [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    let(:job_id) { 'this-is-job' }
+
+    let(:job) { double }
+
+    let(:jobs) { [job] }
+
+    before(:each) do
+      allow(job).to receive(:safe_to_clean_up?).and_return(true)
+      allow(job).to receive(:is_finished?).and_return(false)
+      allow(job).to receive(:id).and_return(job_id)
```
Comment (seanstory) [inline spec/connectors_async/job_watcher_spec.rb:110]: No need to change it for this PR, but just an FYI that you can avoid having all these `allow` statements if you do something like:

```
let(:job) { double('job', :safe_to_clean_up? => true, :is_finished? => false, :id => job_id) }
```

Comment (seanstory) [reply spec/connectors_async/job_watcher_spec.rb:110]: https://apidock.com/rspec/Spec/Mocks/ExampleMethods/double

**Review by seanstory — APPROVED** [review]
A few small comments, but overall LGTM. 

It would be awesome to have you make a small diagram or something, to update the team on how these async classes work together. Or if it's simple enough, it might be worth describing in the README or in a block comment inside `connectors_async.rb` or something.

File: lib/connectors_async/job_store.rb [hunk 3]
```diff
@@ -16,6 +17,8 @@ class JobNotFoundError < StandardError; end
     def initialize
       # multiple threads can write to the store
       @store = Concurrent::Hash.new
+      @watcher = ConnectorsAsync::JobWatcher.new(job_store: self)
```
Comment (wangch079) [inline lib/connectors_async/job_store.rb:20]: If we make `JobStore` and `JobWatcher` singleton, they don't need to reference each other any more.

File: spec/connectors_async/job_watcher_spec.rb (hunk 2 above)
Comment (artem-shelkovnikov) [reply spec/connectors_async/job_watcher_spec.rb:110]: Woah didn't know about it, thanks for sharing! I'll change the code to use this practice too (since I have to change other parts and it'll drop the review anyway)

File: lib/connectors_async/job.rb [hunk 4]
```diff
@@ -78,6 +83,7 @@ def pop_batch(up_to: 50)
       results
     rescue ThreadError
       puts 'Attempt to access an empty queue happened, when the queue was not supposed to be empty'
+      notify_changed
```
[the remark this answers is not in the document]
Comment (artem-shelkovnikov) [reply lib/connectors_async/job.rb:86]: We do, I've missed it.

File: lib/connectors_async/job.rb [hunk 5]
```diff
@@ -86,5 +92,18 @@ def is_finished?
 
       [ConnectorsShared::JobStatus::FINISHED, ConnectorsShared::JobStatus::FAILED].include?(status)
     end
+
+    def safe_to_clean_up?
+      return true if is_finished? && @data[:documents].empty?
```
[the remark this answers is not in the document]
Comment (artem-shelkovnikov) [reply lib/connectors_async/job.rb:97]: That's true, however Enterprise Search also consumes the documents in the queue even from the FAILED job. However, I think for this logic here it might make sense to ignore the state of the queue.

File: lib/app/app.rb [hunk 6]
```diff
@@ -19,7 +19,10 @@
 require 'connectors_shared'
 require 'connectors_app/config'
 require 'connectors_sdk/base/registry'
-require 'connectors_async'
```
[the remark this answers is not in the document]
Comment (artem-shelkovnikov) [reply lib/app/app.rb]: I've reverted the change back, but tbh I like specific requires instead of full namespace ones, just for sake of readability and ease of understanding

File: lib/connectors_async/job.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (artem-shelkovnikov) [reply lib/connectors_async/job.rb:98]: Hmmmm no, actually the job can be FINISHED, but the documents still not collected (state machine logic is a bit awkward here).

We need to make sure that the queue is empty, because Enterprise Search will stop only if both queue is empty and status is finished.

This logic can be improved by making state machine more robust, and we'll approach it in future.

File: lib/connectors_async/job.rb [hunk 7]
```diff
@@ -86,5 +92,18 @@ def is_finished?
...
+
+      # half an hour seems good enough, given that some connectors take 5-10 minutes to respond when throttled
+      Time.now - @data[:last_updated_at] > 60 * 30
```
[the remark this answers is not in the document]
Comment (artem-shelkovnikov) [reply lib/connectors_async/job.rb]: I've moved 60*30 to constants for now, good call.

About the threshold difference - the numbers I've chosen are a bit arbitrary and are just based on current connector that I tested (Sharepoint). As Sean has mentioned, they will actually change later based on our actual experience with the connectors.

How critical do you think the difference of 5 min vs 30 min is here? My line of thought was that it's good enough for the prototype to work - the main metric here is our memory usage and thread usage, as we store our jobs in-memory, so the only reason for job clean up logic is memory reduction + freeing up a thread. 30 min threshold here is for the jobs that are actually stuck - in this case quite likely that these jobs have an empty queue (no memory usage) and just hold a thread (that's the biggest issue). On the other hand we're in tech preview, so we don't need to deal with hanging thread yet - user can customise the thresholds themselves + restart the connector occasionally if it has too many stuck threads running.

File: lib/connectors_async/job.rb [hunk 8]
```diff
@@ -86,5 +92,18 @@ def is_finished?
...
+    end
+
+    private
+
+    def notify_changed
+      @data[:last_updated_at] = Time.now
```
[the remark this answers is not in the document]
Comment (artem-shelkovnikov) [reply lib/connectors_async/job.rb:108]: Good call, flipped @data to be an instance of Concurrent::Map (it's stated to be more performant than Concurrent::Hash - https://www.rubydoc.info/gems/concurrent-ruby/1.0.1/Concurrent/Map)

File: lib/connectors_async/job_store.rb (hunk 3 above)
Comment (artem-shelkovnikov) [reply lib/connectors_async/job_store.rb:20]: They still will have dependencies on each other, just implicit.

I'm considering the singleton change, but haven't be able to approach it yet - hopefully soon though.

File: lib/connectors_async/job_watcher.rb [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+module ConnectorsAsync
+  class JobWatcher
+    IDLE_TIME = 30 # seconds
+
+    class AlreadyWatchingError < StandardError; end
+
+    def initialize(job_store:)
+      @job_store = job_store
+
+      @pool = Concurrent::ThreadPoolExecutor.new(
```
[the remark this answers is not in the document]
Comment (artem-shelkovnikov) [reply lib/connectors_async/job_watcher.rb:20]: > Is a ThreadPool necessary here considering we only need one thread and there is only one task to execute?
No, I will change it to just use one thread, good call.

> When I first thought of how to clean stale jobs, I was thinking of cron job. May I know why you favor a watcher instead of a cron?

Having a cron job means that we will expose the API of jobs outside somehow, and it's probably not something we're willing to do now. Also having a cron means that connectors operations will be more complex - now you just do `make run` and forget, but with cron user will have to maintain the cron entry - install it, refresh it if something changed in cron definition, remove it manually, logs for this will be separated too, so a lot of work.

Also JobWatcher is a temporary abstraction - once we persist jobs we won't need it any more imo.

File: lib/connectors_async/job_store.rb [hunk 10]
```diff
@@ -16,6 +17,8 @@ class JobNotFoundError < StandardError; end
...
+      @watcher.watch
```
Comment (tarekziade) [inline lib/connectors_async/job_store.rb:21]: replace this call by `watch_immediatly: true` in the constructor

File: lib/connectors_async/job_watcher.rb [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+
+    def watch
+      raise AlreadyWatchingError.new('Already watching!') unless @is_watching.make_true
+      ConnectorsShared::Logger.info('Watching after jobs {•̃_•̃}')
+
+      Thread.new do
+        loop do
+          run!
+        rescue StandardError => e
+          ConnectorsShared::ExceptionTracking.log_exception(e)
```
Comment (tarekziade) [inline lib/connectors_async/job_watcher.rb:33]: so we catch the error and log it. and the thread exits. nothing else to do after that?

File: lib/connectors_async/job_watcher.rb [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+      jobs_to_clean_up.each do |job|
+        ConnectorsShared::Logger.debug "Cleaning up #{job.id}"
+        job.fail(ConnectorsAsync::Job::StuckError.new) unless job.is_finished?
+        @job_store.delete_job!(job.id)
+      end
+
+      idle(IDLE_TIME)
+    end
+
+    def idle(timeout)
```
Comment (tarekziade) [inline lib/connectors_async/job_watcher.rb:64]: why this method, use `sleep`

**Review by tarekziade — COMMENTED** [review]
Some nits+ one design issue

The `notify_change` is asking every user of the mapping to call it explicitly on every change -- but it's not enforced so we might get bugs in the future here. Since all you seem to care about is keeping the last modified date for your mapping when it's changed, convert this to an internal mechanism on key updates.

File: lib/connectors_async/job_watcher.rb (hunk 9 above)
[the remark this answers is not in the document]
Comment (artem-shelkovnikov) [reply lib/connectors_async/job_watcher.rb]: Do you mean by cron having a separate record in `crontab` that executes some script to clean up jobs? Can you elaborate more about this?

> Could you elaborate on why it's not needed anymore when jobs are moved from memory to datastore? I think we don't need to clean up jobs when they are persisted. But we will need to clean up the documents queue, which should be separated from Job by then.

What I've meant is that the logic will change heavily, such as the code that cleans up stuff won't be needed much - in the new world we will still have some sort of the queue of course, but it will be detached from everything that's happening in current Job class. This queue will be working differently - there will be a worker that picks up data from it and tries to send to ElasticSearch as fast as possible, but the queue itself should never be cleaned up - it should be persisted in ElasticSearch, thus a case when queue is stuck - not sending documents - is an exception that should probably even terminate the process? This is something that we'll decide, but design of the ingestion process in the future will be just really really different in some parts, and I see the cleanup introduced here as something temporary, changing heavily in upcoming future.

File: lib/connectors_async/job.rb [hunk 13]
```diff
@@ -12,13 +14,14 @@ module ConnectorsAsync
...
     class InvalidStatusError < StandardError; end
+    class StuckError < StandardError; end
 
     def initialize(job_id)
-      @data = {
-        :job_id => job_id,
-        :status => ConnectorsShared::JobStatus::CREATED,
-        :documents => Queue.new
-      }
+      @data = Concurrent::Map.new
```
[the remark this answers is not in the document]
Comment (artem-shelkovnikov) [reply lib/connectors_async/job.rb]: Yeah looking at that again, I'm not sure if we gain anything here by using a Concurrent::Map - we have a limited number of keys that we use in the hash, thus we don't supposedly have a risk when writing into different new keys in parallel threads (that will internally point to the same hash bucket).

File: lib/connectors_async/job_watcher.rb (hunk 9 above)
[the remark this answers is not in the document]
Comment (wangch079) [reply lib/connectors_async/job_watcher.rb]: >Do you mean by cron having a separate record in `crontab` that executes some script to clean up jobs? 

We actually have a similar cron job in `ent-search`, Check [this](https://github.com/elastic/ent-search/blob/main/frito_pie/lib/work/cron/delete_old_connectors_jobs.rb)

>there will be a worker that picks up data from it and tries to send to ElasticSearch as fast as possible

Make sense. Forgot the new ingestion is push model

File: lib/connectors_async/job.rb [hunk 14]
```diff
@@ -86,5 +94,18 @@ def is_finished?
...
+      return true if is_finished? && @data[:documents].empty?
+
+      # half an hour seems good enough, given that some connectors take 5-10 minutes to respond when throttled
+      Time.now - @data[:last_updated_at] > ConnectorsShared::Constants::STALE_JOB_TIMEOUT
+    end
+
+    private
+
+    def notify_changed
+      @data[:last_updated_at] = Time.now
```
[the remark this answers is not in the document]
Comment (artem-shelkovnikov) [reply lib/connectors_async/job.rb]: As discussed, for now I move `:last_updated_at` into its own variable and make sure that we don't expose internal `@data` hash or `@last_updated_at` outside of Job class. I've also added a comment about it in the `initialize` method of Job class.

File: lib/connectors_async/job_watcher.rb (hunk 12 above)
Comment (artem-shelkovnikov) [reply lib/connectors_async/job_watcher.rb:65]: Added a logger.debug inside

File: lib/connectors_async/job_store.rb (hunk 10 above)
Comment (artem-shelkovnikov) [reply lib/connectors_async/job_store.rb:21]: It's surprisingly hard to test if it's done this way.

I also think it's better to keep constructor from having big side effects, such as spawning a new thread - makes it easier to maintain and read the code.

File: lib/connectors_async/job_watcher.rb (hunk 11 above)
Comment (artem-shelkovnikov) [reply lib/connectors_async/job_watcher.rb:34]: I tried to cover it with the following test:
https://github.com/elastic/connectors/blob/57b24179042dba04a46870a585eeee28b4de8fef/spec/connectors_async/job_watcher_spec.rb#L54

In this example first call of `run!` raises a StandardError, and consecutive runs don't raise an error and do nothing. I've also set up `loop` to yield exactly 3 times, while in reality it yields infinitely, and the test checks that it's actually called 3 times.

So tldr - I log an error but the thread keeps running infinitely until the process is killed

**Review by seanstory — APPROVED** [review]
:rubber-stamp:

File: lib/connectors_async/job_watcher.rb (hunk 9 above)
[the remark this answers is not in the document]
Comment (artem-shelkovnikov) [reply lib/connectors_async/job_watcher.rb]: > We actually have a similar cron job in ent-search, Check [this](https://github.com/elastic/ent-search/blob/main/frito_pie/lib/work/cron/delete_old_connectors_jobs.rb)
Oh wow, I totally forgot about these ones - thanks for sharing, it might also be something we could use in future when managing jobs!
